src/service.rs

```diff
@@ -143,86 +142,113 @@ impl<Types: NodeType> GlobalState<Types> {
 
         // update the builder state to last built block
         self.builder_state_to_last_built_block
-            .insert(builder_vid_commitment, response_msg);
+            .insert((builder_vid_commitment, view_num), response_msg);
     }
 
     // remove the builder state handles based on the decide event
     pub fn remove_handles(
         &mut self,
         builder_vid_commitment: &VidCommitment,
-        block_hashes: HashSet<(Types::Time, BuilderCommitment)>,
+        block_hashes: HashSet<(VidCommitment, BuilderCommitment, Types::Time)>,
         on_decide_view: Types::Time,
         bootstrap: bool,
     ) {
         // remove the builder commitment from the spawned builder states
         if !bootstrap {
-            let view_num = self.spawned_builder_states.remove(builder_vid_commitment);
-            if view_num.is_some() {
-                tracing::info!(
-                    "Removing handles for builder view num {:?}",
-                    view_num.unwrap()
-                );
-            }
+            // remove everything from the spawned builder states when view_num <= on_decide_view
+            self.spawned_builder_states
+                .retain(|(_vid, view_num), _channel| view_num > &on_decide_view);
         }
-        {
-            let cleanup_after_view = on_decide_view + self.buffer_view_num_count;
-
-            let edit = self
-                .view_to_cleanup_targets
-                .entry(cleanup_after_view)
-                .or_insert((Default::default(), Default::default()));
-            edit.0.push(*builder_vid_commitment);
-
-            for (view_num, block_hash) in block_hashes {
-                edit.1.push(block_hash.clone());
-                tracing::debug!(
-                    "GC view_num {:?}: block_hash {:?}, deferred to view {:?} ",
-                    view_num,
-                    block_hash,
-                    cleanup_after_view
-                );
-            }
+
+        let cleanup_after_view = on_decide_view + self.buffer_view_num_count;
+
+        let edit = self
+            .view_to_cleanup_targets
+            .entry(cleanup_after_view)
+            .or_insert((Default::default(), Default::default()));
+        edit.0.push(*builder_vid_commitment);
```

Comment:
This would be much clearer if we had a field name instead of `.0`

Reply:
https://github.com/EspressoSystems/hotshot-builder-core/pull/120/commits/08c8e9d95a57d738a240a91c86511a04dd384aa7